src/metabase/lib/schema/expression.cljc

```diff
@@ -1,39 +1,191 @@
 (ns metabase.lib.schema.expression
   (:require
+   [malli.core :as mc]
+   [metabase.lib.schema.common :as common]
    [metabase.lib.schema.filter :as filter]
    [metabase.lib.schema.literal :as literal]
    [metabase.lib.schema.ref :as ref]
+   [metabase.lib.schema.temporal-bucketing :as temporal-bucketing]
+   [metabase.types]
    [metabase.util.malli.registry :as mr]))
 
+(comment metabase.types/keep-me)
+
+(defn- ref-of-base-type [base-type]
+  [:and
+   [:ref ::ref/ref]
+   [:fn
+    {:error/message (str ":field, :expression, :or :aggregation with a " base-type " :base-type")}
+    (fn [[_ref opts]]
+          (isa? (:base-type opts) base-type))]])
+
+(defn- ref-of-base-type-with-temporal-unit [base-type temporal-unit-schema]
+  [:and
+   (ref-of-base-type base-type)
+   [:fn
+    {:error/message (str ":field, :expression, :or :aggregation reference returning a "
+                         base-type
+                         " with a :temporal-unit that matches "
+                         ;; TODO -- humanize this
+                         (pr-str temporal-unit-schema))}
+    (let [validator (mc/validator temporal-unit-schema)]
+      (fn [[_ref {:keys [temporal-unit], :as _opts}]]
+        (validator temporal-unit)))]])
+
 ;;; An expression that we can filter on, or do case statements on, etc.
 (mr/def ::boolean
   [:or
-   ::literal/boolean
-   [:ref ::filter/filter]])
+   {:error/message "expression returning a boolean"}
```

Comment:
It doesn't seem like Malli actually respects this `:error/message`, despite the documentation suggesting it should work everywhere, but I added it anyway and maybe if it gets fixed upstream we'll have nicer error messages.

Reply:
If you mean that passing the wrong things to `mu/defn` doesn't respect this, that'll be because the error reporting in `mu/defn` is custom. Probably easy to add a short-circuit case to the `metabase.util.malli` error humanization to use this labels if present.